test/ecto/adapters/tds_test.exs

```diff
@@ -932,6 +932,30 @@ defmodule Ecto.Adapters.TdsTest do
                ~s{WHERE ((s0.[id] > 0) AND (s0.[id] < @3))}
   end
 
+  test "inner lateral join with fragment" do
+    query = Schema
+            |> join(:inner_lateral, [p], q in fragment("SELECT * FROM schema2 AS s2 WHERE s2.id = ? AND s2.field = ?", p.x, ^10))
+            |> select([p, q], {p.id, q.z})
+            |> where([p], p.id > 0 and p.id < ^100)
+            |> plan()
+    assert all(query) ==
+           ~s{SELECT s0.[id], f1.[z] FROM [schema] AS s0 CROSS APPLY } <>
+           ~s{((SELECT * FROM schema2 AS s2 WHERE s2.id = s0.[x] AND s2.field = @1)) AS f1 } <>
+           ~s{WHERE ((s0.[id] > 0) AND (s0.[id] < @2))}
+  end
+
+  test "left lateral join with fragment" do
+    query = Schema
+            |> join(:left_lateral, [p], q in fragment("SELECT * FROM schema2 AS s2 WHERE s2.id = ? AND s2.field = ?", p.x, ^10))
+            |> select([p, q], {p.id, q.z})
+            |> where([p], p.id > 0 and p.id < ^100)
+            |> plan()
+    assert all(query) ==
+           ~s{SELECT s0.[id], f1.[z] FROM [schema] AS s0 OUTER APPLY } <>
+           ~s{((SELECT * FROM schema2 AS s2 WHERE s2.id = s0.[x] AND s2.field = @1)) AS f1 } <>
```

Comment:
You might notice that there are double brackets here.   I've found a bug with fragments where if a fragment references field from the base schema, it's properly wrapped with brackets () when it's received by join/2 in lib/ecto/adapters/tds/connection.ex.   However, if the fragment references values from another join, the fragment received by join is not pre-wrapped with brackets.   

Since MS SQL Server is forgiving for double brackets (()), my change to connection.ex is safe and it works around the fragment bracket wrapping bug.   I'll try to track down the fragment bracket wrapping bug.  If I can find the proper fix, I'll submit a separate PR and undo the workaround join_expr/2 workaround I introduced in lib/ecto/adapters/tds/connection.ex.

Reply:
It turns out that there was a bug in **lib/ecto/adapters/tds/connection.ex** that resulted in this discrepancy.   See commit fd6321a for the fix.